### PR TITLE
Improved rendering system

### PR DESCRIPTION
**Pull request by blackwarthog**
Also fixes the performance issue with complex vector artwork #314

### CODE REVIEW

File: synfig-core/src/synfig/rendering/software/function/blur.cpp [hunk 1]
```diff
@@ -81,17 +81,23 @@ software::Blur::Params::validate()
...
-	etl::set_intersect(dest_rect, dest_rect, src_rect - src_offset + dest_rect.get_min());
+
+	dest_rect = src_rect - offset;
+	dest_rect.minx += extra_size[0];
+	dest_rect.miny += extra_size[1];
+	dest_rect.maxx -= extra_size[0];
+	dest_rect.maxy -= extra_size[1];
 	if (!dest_rect.valid()) return false;
-	offset = src_offset - src_rect.get_min();
+	if (!etl::contains(RectInt(0, 0, dest->get_w(), dest->get_h()), dest_rect)) return false;
```
Comment (ice0) [inline synfig-core/src/synfig/rendering/software/function/blur.cpp]: return etl::contains(RectInt(0, 0, dest->get_w(), dest->get_h()), dest_rect);

File: synfig-core/src/synfig/matrix.h [hunk 2]
```diff
@@ -365,9 +365,14 @@ class Matrix3
 	Matrix3 operator+(const Matrix3 &rhs)const
 		{ return Matrix3(*this) += rhs; }
 
+	value_type det() const;
```
Comment (ice0) [inline synfig-core/src/synfig/matrix.h]: I don't understand what 'det' is mean

File: synfig-core/src/synfig/rendering/software/surfacesw.cpp [hunk 3]
```diff
@@ -50,84 +50,106 @@ using namespace rendering;
...
+	return &(*this->surface)[0][0];
+}
+
 void
 SurfaceSW::set_surface(synfig::Surface &surface, bool own_surface)
 {
-	this->own_surface = own_surface;
-
-	if (&surface == this->surface)
+	if (&surface == this->surface) {
```
Comment (ice0) [inline synfig-core/src/synfig/rendering/software/surfacesw.cpp]: if surface and this->surface both equals NULL is this ok?

File: synfig-core/src/synfig/rendering/software/surfacesw.cpp [hunk 4]
```diff
@@ -50,84 +50,106 @@ using namespace rendering;
...
+		assert(this->surface);
+		delete(this->surface);
+	}
 
 	this->surface = &surface;
 	assert(this->surface);
-	mark_as_created(false);
-	set_size(this->surface->get_w(), this->surface->get_h());
-	mark_as_created(this->surface->get_w() > 0 && this->surface->get_h() > 0);
+	set_desc(surface.get_w(), surface.get_h(), false);
```
Comment (ice0) [inline synfig-core/src/synfig/rendering/software/surfacesw.cpp]: Here is unclear why 'set_desc'->set_description? got width, height and false

File: synfig-core/src/synfig/rendering/software/surfacesw.h [hunk 5]
```diff
@@ -46,28 +46,35 @@ class SurfaceSW: public Surface
...
-	virtual bool get_pixels_vfunc(Color *buffer) const;
+	virtual bool clear_vfunc();
+	virtual bool reset_vfunc();
+	virtual const Color* get_pixels_pointer_vfunc() const;
 
 public:
 	SurfaceSW();
-	explicit SurfaceSW(const Surface &other);
+	explicit SurfaceSW(const Surface &other, bool own_surface);
 	~SurfaceSW();
```
Comment (ice0) [inline synfig-core/src/synfig/rendering/software/surfacesw.h]: Didn't see here surface width and height, and it is unclear where i can get them

File: synfig-core/src/synfig/rendering/surface.cpp [hunk 6]
```diff
@@ -52,98 +48,247 @@ using namespace rendering;
...
+bool
+Surface::get_pixels_vfunc(Color *dest) const
 {
-	assert(!is_created());
-	unset_alternative();
-	this->width = width > 0 ? width : 0;
-	this->height = height > 0 ? height : 0;
+	const Color *src = get_pixels_pointer();
+	if (!src) return false;
+	memcpy(dest, src, get_buffer_size());
```
Comment (ice0) [inline synfig-core/src/synfig/rendering/surface.cpp]: Is dest has enough size for src? No checks, and also unclear

File: synfig-core/src/synfig/rendering/surface.h [hunk 7]
```diff
@@ -43,53 +46,206 @@ namespace synfig
...
+
+	virtual bool create_vfunc(int width, int height)
+		{ return false; }
+	virtual bool assign_vfunc(const Surface &other)
+		{ return false; }
+	virtual bool clear_vfunc()
+		{ return false; }
+	virtual bool reset_vfunc()
+		{ return false; }
+	//! Implementations of this function should to work quick
```
Comment (ice0) [inline synfig-core/src/synfig/rendering/surface.h]: Implementations of this function **MUST** work fast

**Comment by blackwarthog** [comment]
> I don't understand what 'det' is mean

Determinant
https://en.wikipedia.org/wiki/Determinant

**Comment by blackwarthog** [comment]
> >  	if (!dest_rect.valid()) return false;
> -	offset = src_offset - src_rect.get_min();
> +	if (!etl::contains(RectInt(0, 0, dest->get_w(), dest->get_h()), dest_rect)) return false;
> 
> return etl::contains(RectInt(0, 0, dest->get_w(), dest->get_h()), dest_rect);

In this function i've used construction `if (all-bad) return false;` several times. I think be batter to keep this style at the end, it looks more clear.

**Comment by blackwarthog** [comment]
> It is unclear what is the task of this class 

Where?

**Pull request merged**